linera-witty/src/type_traits/implementations/std/tuples.rs

```diff
@@ -5,12 +5,35 @@
 
 use crate::{
     GuestPointer, InstanceWithMemory, Layout, Memory, Runtime, RuntimeError, RuntimeMemory,
-    WitLoad, WitType,
+    WitLoad, WitStore, WitType,
 };
-use frunk::{hlist_pat, HList};
+use frunk::{hlist, hlist_pat, HList};
 
+/// Implement [`WitType`], [`WitLoad`] and [`WitStore`].
+///
+/// When implementing [`WitStore`] for tuples, it's necessary to deconstruct the tuple and rebuild
+/// it as an heterogeneous list. However, because the methods receive `&self`, the deconstruction
+/// leads to references to the elements. Therefore an extra constraint is necessary, which is that
+/// the reference also implements [`WitStore`] and that the layout is the same as the referenced
+/// type.
+///
+/// Using this clause for the unit type leads to a compiler error, because it tries to match the
+/// layout type to itself, which the compiler doesn't handle correctly. The solution to this is to
+/// only add the clause for the implementations that have elements.
 macro_rules! impl_wit_traits {
+    () => {
+        impl_wit_traits!(@generate;);
+    };
+
     ($( $names:ident : $types:ident ),*) => {
+        impl_wit_traits!(
+            @generate $( $names: $types ),* ;
+            for<'a> HList![$( &'a $types ),*]:
+                WitType<Layout = <HList![$( $types ),*] as WitType>::Layout> + WitStore,
+        );
+    };
+
+    (@generate $( $names:ident : $types:ident ),* ; $( $borrow_store_clause:tt )*) => {
```

Comment:
Ideally remove `@generate` as we did before?

Reply:
Done, thanks!